pkg/kubelet/cm/container_manager_linux.go

```diff
@@ -614,77 +617,62 @@ func (cm *containerManagerImpl) setFsCapacity() error {
 }
 
 // TODO: move the GetResources logic to PodContainerManager.
-func (cm *containerManagerImpl) GetResources(pod *v1.Pod, container *v1.Container, activePods []*v1.Pod) (*kubecontainer.RunContainerOptions, error) {
-	opts := &kubecontainer.RunContainerOptions{}
-	// Gets devices, mounts, and envs from device plugin handler.
-	glog.V(3).Infof("Calling devicePluginHandler AllocateDevices")
-	// Maps to detect duplicate settings.
-	devsMap := make(map[string]string)
-	mountsMap := make(map[string]string)
-	envsMap := make(map[string]string)
-	allocResps, err := cm.devicePluginHandler.Allocate(pod, container, activePods)
-	if err != nil {
-		return opts, err
-	}
-	// Loops through AllocationResponses of all required extended resources.
-	for _, resp := range allocResps {
-		// Each Allocate response has the following artifacts.
-		// Environment variables
-		// Mount points
-		// Device files
-		// These artifacts are per resource per container.
-		// Updates RunContainerOptions.Envs.
-		for k, v := range resp.Envs {
-			if e, ok := envsMap[k]; ok {
-				glog.V(3).Infof("skip existing envs %s %s", k, v)
-				if e != v {
-					glog.Errorf("Environment variable %s has conflicting setting: %s and %s", k, e, v)
-				}
-				continue
-			}
-			envsMap[k] = v
-			opts.Envs = append(opts.Envs, kubecontainer.EnvVar{Name: k, Value: v})
+func (cm *containerManagerImpl) GetResources(pod *v1.Pod, container *v1.Container) (*kubecontainer.RunContainerOptions, error) {
+	// Allocate is thought to be called during predicate phase,
+	// just try to fetch device runtime information from cached state here
+	devOpts := cm.devicePluginHandler.GetDeviceRunContainerOptions(pod, container)
+	if devOpts == nil {
+		return nil, fmt.Errorf("Failed to get device run container options")
+	}
+	opts := &kubecontainer.RunContainerOptions{
+		Devices: devOpts.Devices,
+		Mounts:  devOpts.Mounts,
+		Envs:    devOpts.Envs,
+	}
+
+	return opts, nil
+}
+
+func (cm *containerManagerImpl) UpdatePluginResources(node *schedulercache.NodeInfo, attrs *lifecycle.PodAdmitAttributes) error {
+	pod := attrs.Pod
```

Comment:
Instead of putting the functionality in this cm function, the code structure is probably cleaner if the device plugin related functionality is moved to device plugin handler. I think we can change devicePluginHandler.Allocate() to take pod and nodeInfo as parameters, and scan through pod.Spec.Containers there. The existing devicePluginHandler.Allocate() can be turned into an internal function, say allocateContainerResources().